target_decisioning_engine/tests/test_artifact_provider.py

```diff
@@ -0,0 +1,257 @@
+            self.provider = ArtifactProvider(config)
+            self.provider.initialize()
+            self.assertEqual(mock_fetch_artifact.call_count, 1)
+            self.assertEqual(self.provider.artifact_location,
+                             "https://assets.adobetarget.com/client123/production/v1/rules.json")
+            self.assertTrue(self.provider.polling_timer.is_alive())
+
+    def test_initialize_artifact_location_is_string(self):
+        config = DecisioningConfig("client123", "org999", polling_interval=350, artifact_location="Mordor")
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None) as mock_fetch_artifact:
+            self.provider = ArtifactProvider(config)
+            self.provider.initialize()
+            self.assertEqual(mock_fetch_artifact.call_count, 1)
+            self.assertEqual(self.provider.artifact_location, "Mordor")
+            self.assertTrue(self.provider.polling_timer.is_alive())
+
+    def test_initialize_artifact_from_config(self):
+        config = DecisioningConfig("client123", "org999", polling_interval=350, artifact_location="Mordor",
+                                   artifact_payload={"a": 1, "b": 2})
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None) as mock_fetch_artifact:
+            self.provider = ArtifactProvider(config)
+            self.provider.initialize()
+            mock_fetch_artifact.assert_not_called()
+            self.assertEqual(self.provider.artifact_location, "Mordor")
+            self.assertEqual(self.provider.artifact, {"a": 1, "b": 2})
+            self.assertTrue(self.provider.polling_timer.is_alive())
+
+    def test_add_remove_subscription(self):
+        def subscriber():
+            pass
+
+        self.provider = ArtifactProvider(self.default_config)
+
+        # Test add_subscription
+        self.assertEqual(self.provider.subscription_count, 0)
+        self.provider.add_subscription(subscriber)
+        self.assertEqual(self.provider.subscription_count, 1)
+        self.assertEqual(self.provider.subscriptions[1], subscriber)
+
+        # Test remove_subscription
+        self.provider.remove_subscription(1)
+        self.assertEqual(self.provider.subscription_count, 0)
+
+    def test_remove_subscription_bad_key(self):
+        self.provider = ArtifactProvider(self.default_config)
+        self.assertEqual(self.provider.subscription_count, 0)
+        self.provider.remove_subscription(100)
+        self.assertEqual(self.provider.subscription_count, 0)
+
+    def test_emit_new_artifact(self):
+        emitter = Mock()
+        subscriber = Mock()
+        config = DecisioningConfig("client123", "org999", event_emitter=emitter)
+        self.provider = ArtifactProvider(config)
+        self.provider.add_subscription(subscriber)
+        self.assertEqual(self.provider.subscription_count, 1)
+
+        payload = {"a": 1, "b": 2}
+        self.provider.emit_new_artifact(payload)
```

Comment:
This may validate that the `emit_new_artifact` emits the proper event.  But you really need to validate that ArtifactProvider calls this method on its own after retrieving the artifact.  Otherwise, you get a false sense of security with this test.